### PR TITLE
Fix order local refit copies after source updates

### PR DESCRIPTION
**Pull request by Sunt-ing**
- [X] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

Local same-rank refit copies can silently read stale model weights when the source update is still queued on the caller's CUDA stream. NCCL, Gloo, NIXL, and NVSHMEM all run the local copy on a dedicated stream, but only order the caller stream after that copy; they do not order the copy stream after the producer before it reads the source tensor.

This change makes each local copy stream wait for the caller's current stream before copying. It preserves the existing copy-to-caller join and does not change the public API or remote-rank transport.

## Tests

```bash
python -m torch.distributed.run --nproc-per-node 8 -m pytest -q \
  tests/unit_tests/resharding/test_copy_services.py::test_local_copy_waits_for_current_stream \
  tests/unit_tests/resharding/test_copy_services.py::test_optional_backend_local_copy_waits_for_current_stream
```

With the fix, every rank passes all four backend cases (`4 passed` per rank). Removing only the corresponding `wait_stream` calls makes every affected case fail with the destination retaining the stale value (maximum absolute difference `1.0`). The NIXL/NVSHMEM test uses real CUDA streams and tensors while mocking only the optional remote transport, so it does not require those optional runtime dependencies in the unit-test environment.

As supplemental validation, I also exercised the public `swap_model_weights()` path after fused AdamW updates on a 20-layer, hidden-size 4096 GPT model. The source/destination logits were the oracle:

| Backend | Current code | Fixed | Exact revert |
| --- | ---: | ---: | ---: |
| NCCL | max diff `3.0010` | `0` | max diff `3.3286` |
| Gloo | max diff `2.8388` | `0` | max diff `2.8373` |
| NIXL | max diff `0.9583` | `0` | max diff `7.2628` |
| NVSHMEM | max diff `0.9849` | `0` | max diff `0.9629` |

### NCCL/Gloo refit latency

I measured the cost of the added dependency with the repository's `examples/rl/benchmark_refit.py` on one H200. Both variants used collocated TP1/PP1/EP1, BF16, a 20-layer/4096-hidden/32-head GPT model (4.23B parameters per model), sequence length 8, batch size 1, five warmup iterations, and 30 timed iterations. Each timed iteration synchronizes CUDA before refit, so this comparison measures the dependency overhead when source work is already complete.

The fixed tree differed from the current tree only by this PR. The same temporary current-main API compatibility adjustment and statistics output were applied to both trees and are not part of the PR. All values below are milliseconds; `mean / median / stddev / min-max` summarizes all 30 samples.

| Backend |                                Current |                                  Fixed |           Median delta |             Mean delta |
| ------- | -------------------------------------: | -------------------------------------: | ---------------------: | ---------------------: |
| NCCL    |     `7.39 / 7.38 / 0.05 / 7.33-7.52` |     `7.51 / 7.45 / 0.24 / 7.40-8.67` | `+0.07` (`+0.95%`) | `+0.12` (`+1.62%`) |
| Gloo    | `16.34 / 8.28 / 44.07 / 8.24-249.67` | `16.21 / 8.27 / 43.47 / 8.24-246.39` | `-0.01` (`-0.12%`) | `-0.13` (`-0.80%`) |

Each Gloo arm contained one approximately 250 ms outlier, which inflates its mean and standard deviation; the median shows no measurable regression at the reported precision. If source work is still pending, the new dependency intentionally includes the time required before those weights are valid rather than allowing a stale copy to return early.

### Pre-checks

- [X] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [X] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### CODE REVIEW

**Comment by YangFei1990** [comment]
Hi @Sunt-ing, could you check the comments from agent?

**Comment by Sunt-ing** [comment]
@YangFei1990 Sure. 

First, thanks for flagging NIXL and NVSHMEM. I reproduced the same bug in both backends through the public `swap_model_weights()` path after fused AdamW updates on a 4.23B GPT model. At the current PR head, the destination logits were stale, with maximum differences of `0.9583` for NIXL and `0.9849` for NVSHMEM. Adding one producer-to-copy stream dependency to each backend made all three iterations exact. Removing just those two lines brought the stale results back.

I extended the patch and CUDA ordering test to both optional backends. The test uses real CUDA streams and tensors and mocks only the remote transport, so it does not require the optional NIXL or NVSHMEM runtime. Under the repository's 8-GPU `torchrun` topology, all four backend cases pass on every rank. Reverting only the NIXL and NVSHMEM waits makes both cases fail on every rank, with the destination retaining `0` and a maximum difference of `1.0`. The gate now uses `get_cycles_per_ms()`, and the expected all-ones tensor is independent of the source.

Each wait remains at its backend's local-copy site, one line next to the stream it orders. A shared helper would add indirection while all four backends would still need separate calls. I did not add a Gloo cleanup fixture because `GlooCopyService` inherits the no-op `CopyService.close()`, so the fixture would not release the process group. The test creates the group collectively on every rank in the same `torchrun` invocation.

**Comment by kvareddy** [comment]
@wdykas can you please take a look at this MR? Also, let's create a review group for refit related code.

**Review by wdykas — APPROVED** [review]
Thanks for the contribution! Nice find, thanks for updating all the colocated backends. @Sunt-ing what are you using this for?

**Comment by wdykas** [comment]
/ok to test [aee431b](https://github.com/NVIDIA/Megatron-LM/pull/6823/commits/aee431bfa8824d15dc98ce786c768b6858c96d19)

**Comment by Sunt-ing** [comment]
@wdykas  Thanks for the approval! I'm evaluating it for collocated RL training, where the rollout model needs the latest trainer weights after each optimizer step.

**Review by kvareddy — APPROVED** [review]

**Review by shanmugamr1992 — APPROVED** [review]
